extended/src/main/java/io/kubernetes/client/extended/kubectl/KubectlRollout.java

```diff
@@ -308,4 +330,241 @@ public V1PodTemplateSpec execute() throws KubectlException {
+            add(AnnotationKeyConstants.LAST_APPLIED_CONFIG_ANNOTATION);
+            add(DeploymentHelper.REVISION_ANNOTATION);
+            add(DeploymentHelper.REVISION_HISTORY_ANNOTATION);
+            add(DeploymentHelper.DESIRED_REPLICAS_ANNOTATION);
+            add(DeploymentHelper.MAX_REPLICAS_ANNOTATION);
+          }
+        };
+
+    KubectlRolloutUndo() {
+      super(KubectlRollout.this.apiTypeClass);
+      toRevision = 0L;
+    }
+
+    public KubectlRolloutUndo toRevision(long toRevision) {
+      this.toRevision = toRevision;
+      return this;
+    }
+
+    private long toRevision;
+
+    private void validate() throws KubectlException {
+      StringBuilder msg = new StringBuilder();
+      if (name == null) {
+        msg.append("Missing name, ");
+      }
+      if (namespace == null) {
+        msg.append("Missing namespace, ");
+      }
+      if (toRevision < 0) {
+        msg.append("ToRevision must be a positive integer: ").append(toRevision);
+      }
+      if (msg.length() > 0) {
+        throw new KubectlException(msg.toString());
+      }
+    }
+
+    @Override
+    public ApiType execute() throws KubectlException {
+      validate();
+      AppsV1Api api = new AppsV1Api(this.apiClient);
+      refreshDiscovery();
+      try {
+        if (apiTypeClass.equals(V1Deployment.class)) {
+          V1Deployment deployment = api.readNamespacedDeployment(name, namespace, null, null, null);
+          return deploymentRollBack(deployment, api);
+        } else if (apiTypeClass.equals(V1StatefulSet.class)) {
+          V1StatefulSet statefulSet =
+              api.readNamespacedStatefulSet(name, namespace, null, null, null);
+          return statefulSetRollBack(statefulSet, api);
+        } else {
+          throw new KubectlException("Unsupported class for rollout history: " + apiTypeClass);
+        }
+      } catch (ApiException ex) {
+        throw new KubectlException(ex);
+      }
+    }
+
+    private ApiType deploymentRollBack(V1Deployment deployment, AppsV1Api api)
+        throws ApiException, KubectlException {
+      V1ReplicaSet rsForRevision = deploymentRevision(deployment, api);
+      Boolean paused = deployment.getSpec().getPaused();
+      if (paused != null && paused) {
+        throw new ApiException(
+            "Cannot rollback a paused deployment, resume it first and try again");
+      }
+      if (DeploymentHelper.equalIgnoreHash(
+          rsForRevision.getSpec().getTemplate(), deployment.getSpec().getTemplate())) {
+        logger.info("Current template already matches revision {}", toRevision);
+        return (ApiType) deployment;
+      }
+
+      rsForRevision
+          .getSpec()
+          .getTemplate()
+          .getMetadata()
+          .getLabels()
+          .remove(DeploymentHelper.DEFAULT_DEPLOYMENT_UNIQUE_LABEL_KEY);
+      Map<String, String> annotations = new HashMap<>();
+
+      Map<String, String> deploymentAnnotations = deployment.getMetadata().getAnnotations();
```

Comment:
I feel like this algorithm is complicated. It's easier to do:

```java

Map<String, String> annotations = new Map<>();
annotations.addAll(deployment.getMetadata().getAnnotations());
annotations.addAll(rsForRevision.getMetadata().getAnnotations());
for (String key : annotationsToSkip.keySet()) {
   annotations.remove(key);
}
```

Reply:
I don't think the two  algorithms are equivalent.For keys in "annotationsToSkip", we should keep key/value in "annotations" consistent with deployment.getMetadata().getAnnotations() instead of removing them.

Reply:
Sorry, I mis-read the algorithm. How about 

```java
Map<String, String> annotations = new Map<>();
annotations.addAll(rsForRevision.getMetadata().getAnnotations());
for (String key : annotationsToSkip.keySet()) {
   if (deploymentAnnotations.contains(key)) {
      annotations.put(key, deploymentAnnotations.get(key));
   } else {
     annotations.remove(key);
   }
}
```

I think that is equivalent (add everything from rs, except the ones to skip)